[WIP] [ENHANCEMENT] Arbitrary Equipment Slots

cclark25 opened the pull request:
This PR contains WIP changes to enhance Equipment Slot customization, and I wanted to share it here to gauge interest.

The inspiration here is to modernize equipment controls. With the vanilla C Button item slots, the player has to dedicate 3 buttons to use 3 different items. This was improved with the DPad item slots enhancement, but that is an additional 4 buttons than need to be dedicated.
This is clunky on modern controllers IMO. Ideally it would be much more intuitive to use the same button(s) to use all/some items. Example: Bows, the Hookshot, and the Camera work best on a trigger button (such as R or L).
In 2ship especially, the addition of masks makes the vanilla equipment slots even more strained than in SoH, because the masks require an entire equipment slot for items that are largely just toggle switches and really don't need dedicated buttons for accessibility.
In addition to increasing the number of equipment slots, this also will free up buttons on the controller for other usages, such as splitting out `A` button actions into separate controls.

Anticipated features:
- Customizable number of equipment slots to allow in addition to the C Buttons and DPad equips.
- Alternative equipment slot controller logic.
  - Carousel Logic: Assign an arbitrary number of items to a single button, and use DPad Left/Right (or other buttons) to swap between them.
  - Single Button Item Logic: Items can be assigned to dedicated buttons to switch to them, but they can all be used using the same button.
  - Equipment Slot Item Restrictions: Equipment slots can be restricted to certain items. Examples: Only Masks, Only Bow and Hookshot, etc.
- The new Arbitrary Equipment Slots logic is implemented with a more centralized object-encapsulated design, allowing modders to more easily create and plug-in new equipment logic options in addition to the pre-existing and new options.
- Equipment Slot Display Customization Options: 
  - RGB color selection.
  - Alpha and visibility control.
  - (CODE-MODE ONLY) Animated screen location (used by Carousel).
  - Custom icon size.
  - (POSSIBLE) Custom Slot Icon Loading (currently just defaults to a C Button icon).

This branch was branched off of another enhancement branch I was working on, which was separating certain control events into separate mappings. These changes include separating some `A` button controls into separate mappable buttons.
If this Arbitrary Equipment Slots Enhancement gets a lot of attention and demand, I may rebase to fast track this enhancement.

<!--- section:artifacts:start -->
### Build Artifacts
  - [2ship-linux.zip](https://nightly.link/HarbourMasters/2ship2harkinian/actions/artifacts/1774158439.zip)
  - [2ship-mac.zip](https://nightly.link/HarbourMasters/2ship2harkinian/actions/artifacts/1774188818.zip)
<!--- section:artifacts:end -->

cclark25 committed:
Added initial basic support for intent-based controls.

cclark25 committed:
Added support for cur, prev, press, and release logic.

cclark25 committed:
Updated the intent controls checks.

cclark25 committed:
Altered the bow/hookshot controls so that they can be fired with an arbitrary button.

cclark25 committed:
Added code for swapping active c-buttons.

cclark25 committed:
Began writing code to add in arbitrary equipment buttons

cclark25 committed:
Fixed the arbitrary equipment assignments.

cclark25 committed:
Added ArbitraryItemSlots.

cclark25 committed:
Added an abstraction layer for ArbitraryItemSlotLister so that different implementations of arbitrary item slots can be swapped out at runtime (TODO).

cclark25 committed:
Removed unneeded code.

cclark25 commented:
As-of commit cc82799a27b76d164a7a28d8df026fe9a351ca83, this enhancement successfully implements a prototype of the base functionality.
It adds 2 dedicated equipment slot buttons (recommended to use the L and R buttons).
The slots are red and blue, and currently move side-to-side at the top of the screen to demonstrate animation functionality that can be used in a Carousel option which I have yet to implement.
Items and Masks an both be assigned to these buttons, however non-transformation masks are currently unusable because of a bug that immediately deactivates the mask after activating it.
The custom item slots can also be used during item trade dialogues, though I have only tested this with the Deku Scrubs land deed trades so far.
The item assignments in the pause menu do not however account for swapping logic when the item is assigned to a different button, so the same item can be equipped to multiple slots.
The pause menu assignment animations are also broken currently, but this is mostly just cosmetic.

Caladius commented:
Would be beneficial for gauging interest if you add some screenshots and, preferably, a video of it in action to the GitHub PR. Big fan of modernizing controls especially when it adds additional features.

cclark25 commented:
@Caladius I am having trouble getting a video small enough to upload to github here, so I will have to just share a screenshot for now.

<img width="1440" alt="Screenshot 2024-06-29 at 11 38 00 PM" src="https://github.com/HarbourMasters/2ship2harkinian/assets/38185583/018f3c68-e93b-4d90-907c-dc421969d1b8">


In this screenshot, I am showing that there are 2 additional equip slots which use the L and R buttons, and currently have the Deku Stick and Deku Nuts equipped, respectively.
You can also see that the Deku Stick is currently held by Link, showing that the button works.

What you can't see in this screenshot is that the red and blue item slots are actually moving back and forth from left to right to demonstrate an animated position. This will be very useful when I make the Carousel-style item slot, as the currently active item slot will slide into place when swapping with the Dpad Left/Right buttons.

cclark25 commented:
Okay, I managed to get a video recorded demonstrating the current functionality.

https://github.com/HarbourMasters/2ship2harkinian/assets/38185583/b3d4af3e-b7cb-4ae3-a37f-360f36adbc3e

cclark25 committed:
Added an item carousel item slot manager.

cclark25 commented:
I created a new implementation for arbitrary item slots that uses a carousel to switch between items.
It uses the R button to assign/use the item and the Dpad Left/Right to swap between each slot, with an animated rotation.

https://github.com/HarbourMasters/2ship2harkinian/assets/38185583/adb5d8c0-2ae1-4b9a-8e3b-493fb595a57c

cclark25 committed:
Added some configuration window for item carousels.

cclark25 committed:
Added change to allow the user to configure the number of equipment slots.

cclark25 commented:
I have added some configuration options to an enhancement menu to control the location, color, and number of item slots in the carousel.

https://github.com/HarbourMasters/2ship2harkinian/assets/38185583/b35fdc95-829b-4b86-bf28-9163eeac33db

cclark25 committed:
Fixed the mask activation not working.

cclark25 committed:
Removed an unneeded change.

cclark25 committed:
Fixed the equipment slots so that they respect disabled conditions in non-special cases. Special cases like minigames may still unrestrict the items.

cclark25 committed:
Cleaned up some of the carousel logic so that when unpaused the carousel will automatically move to the next non-disabled slot.

cclark25 committed:
Fixed the Lens of Truth on the arbitrary slots, and corrected some carousel slot swapping logic when the pause menu is closed.

cclark25 committed:
Added an option to change the carousel direction.

cclark25 committed:
Added a number of cosmetic changes to the carousel.

cclark25 committed:
Modified the Carousel Config menu to allow the mapping of the carousel controls.

cclark25 committed:
Added some configuration menu options to use multiple simultaneous configurations.

cclark25 committed:
Added some configuratio options for scale, color, and transparency of arbitrary buttons and fixed a bug with conflicting slot IDs.

cclark25 committed:
Added options for controlling different states of the item slots.

cclark25 committed:
Adjusted the Carousel to use a more configurable set of state options.

cclark25 committed:
Added callbacks to adjust the HUD alphas mirroring the vanilla code, so now arbitrary slots fade off the screen in the same way that normal C slots do on day changes and etc.

cclark25 committed:
Refactored the carousel drawing code to simplify it.

cclark25 commented:
This enhancement is looking pretty good so far.
Right now, I am focussing on getting the arbitrary equipment slots configurations saving, and adding the assigned equipment to the save file.

There is currently an issue I found where bottles are completely broken. A slot with a bottle assigned will never update the contents of the assigned bottle, so they cannot be used to capture bugs, fish, or fairies, and when releasing a bug, fish, or fairy they don't empty the bottle's contents, essentially producing an infinite supply of whatever entity was in the bottle. The equipment slot also doesn't update with new contents if the bottle is updated elsewhere, such as from the save editor, another equipment slot, or buying an item in a store.

There are a number of other minor issues as well:
1. There is no item-swap behavior when assigning an item to a slot when that item is already assigned to another slot. 
2. The item assignment animation does not animate to the correct screen location.
3. Items assigned to arbitrary item slots are not highlighted in the pause menu.
4. There is a possibility of weird visibility in certain HUD conditions in the game, such as in minigames. Not really sure where/how those might look without playing the entire game through.
5. There isn't a way to hide the vanilla slots or control the count of each type of arbitrary equipment slots in the configuration window.

cclark25 committed:
Added functionally to save/load arbitrary equipment slots assigned items to the save file.

cclark25 committed:
Updated the arbitrary slots to save/load configurations to the 2ship2harkinian.json file.

cclark25 commented:
https://youtu.be/O5qVNbQ2ZKE

The 10 MB file size limit on Github is too limiting to properly demonstrate the functionality via video, so here is a Youtube link with the demos for the newest changes.

Since the last video, I have fixed a number of issues the core item slot behavior.
I have also polished the carousel a good bit, and added a configuration menu for customizing the layout and display options.
The config options also persist across executions of the 2ship app, and the items assigned to the arbitrary item slots and carousel are saved to the save file.

The biggest remaining issue is that the bottles don't still work, as I mentioned previously. Masks do work however.
Aside from that issue, these changes are currently in a playable state.

cclark25 commented:
I also wanna say that feedback is welcome right now if anyone has thoughts on the implementation.

cclark25 committed:
Fixed the bottles.

cclark25 commented:
Okay I did some refactoring and now the arbitrary equipment slots now respect bottle contents.
Using a bottled item now consumes the item from the inventory and catching bugs/fish now work as expected.

cclark25 committed:
Began adding logic to add multiple carousels.

cclark25 commented:
I'm working on some logic to allow the user to configure the number of dedicated slots and carousels.
I intended to have this functionality from the start, as I envisioned 2 carousel slots on the screen: 1 using the Left/Right DPad controls and another using the Up/Down DPad controls.
This logic will also serve to disable the 2 arbitrary equipment slot types as well, by defaulting them to 0.

This multiple configuration flow is still a WIP. 

I should also add that these flexible options for equipment slot configurations is heavily dependent on the Intent Control changes I made in Libultraship previously. Without those changes, it would be much more difficult to set up controls for these custom slots, and the number of slots would be limited to a static number for sure.
I'm not really sure that this Enhancement is even feasible without some mod to Libultraship increasing the number of isolated controls available, which is what my Intent Controls mod achieves currently.

cclark25 committed:
Merge remote-tracking branch '2ship/develop' into arbitrary_equipment_buttons

cclark25 committed:
Merge remote-tracking branch '2ship/develop' into arbitrary_equipment_buttons

cclark25 committed:
Ran clang formatter.

cclark25 committed:
Reran clang formatter

cclark25 committed:
Added '-Wno-int-conversion' compiler option to fix compilation issues on my MacBook.

cclark25 committed:
Changed a line of code to hopefully fix failing linux/windows builds.

cclark25 committed:
Made a change to libultraship to hopefully fix the windows build.

cclark25 committed:
Fixed formatter issues and removed some commented code.

cclark25 committed:
Adjusted formatting.

cclark25 committed:
Made some options to improve the number of carousels that can be added.

cclark25 committed:
Fixed the Carousel configuration saving logic so that it saves and loads all the states instead of just the default state.

garrettjoecox commented:
Hey @cclark25,

Have an update for this PR. First off, I want to say that what you put together here looks really awesome. I can tell a lot of work went into it, and I appreciate the effort.

That said, we’ve struggled with how to approach this one. It introduces some patterns that don’t quite align with the rest of the project, and ultimately, it feels more like something that would be a great standalone mod rather than something we’d merge into the base repo. I’m really sorry this feedback is coming so late—it definitely deserved more attention earlier.

If you intend on supporting this for public use I suggest re-opening a PR on your fork for CI builds, and creating a thread in the #2s2h-modding forum on discord!

cclark25 commented:
Hey @garrettjoecox,

No worries. I understand that this is a rather ambitious change and beyond the scope of the core project. Thank you for looking at this PR though!

As it currently stands, I have largely lost a lot of steam for this mod due to lack of feedback here on github and also due to losing access to the discord server.

I do hope that future mod developers could use this branch as a reference for use-cases involving similar functionality, but I likely won't be able to invest the time to maintain this mod except for my own personal usage, though anyone is free to try and use it if they find it here.
